lua/detour/util.lua

```diff
@@ -64,4 +72,42 @@ function M.overlap(window_a, window_b)
     return overlap_helper({M.get_text_area_dimensions(window_a)}, {M.get_text_area_dimensions(window_b)})
 end
 
+function M.find_top_popup()
+    local window_id = vim.api.nvim_get_current_win()
+    local all_covered_windows = vim.tbl_flatten(vim.tbl_values(internal.popup_to_covered_windows))
+    for popup, _ in pairs(internal.popup_to_covered_windows) do
+        if not vim.tbl_contains(all_covered_windows, popup) -- ignore popups with popups nested in them
+            and vim.tbl_contains(M.find_covered_bases(popup), window_id) then
+            return popup
+        end
+    end
+    return window_id -- no popup that covers the current window was found
+end
+
+-- Finds all base windows that are covered by the provided popup.
+-- If the provided window is not a popup, returns the given argument.
+function M.find_covered_bases(window_id)
```

Comment:
Consider just iterating through the base windows in winlayout() and finding the ones the window overlaps with.